### PR TITLE
Fix autosign file permission check [develop]

### PR DESCRIPTION
**Pull request by benediktwerner**
### What does this PR do?
Fixes autosign file permission check `salt.daemons.masterapi.AutoKey._check_permissions()`.

### What issues does this PR fix or reference?
There seems to have been a similar problem in #1870 back in 2012, but the resulting commit 
a8b302b doesn't look much better to me.

### Previous Behavior
The `salt.daemons.masterapi.AutoKey._check_permissions()` function used to check the permissions of an autosign file does some seemingly arbitrary file checks and some of those even twice to check if an `autosign_file` is secure enough to use.

As far as I can tell a file is allowed if **any** of these apply:
- salt is `root` and owns the file
- salt is `root` but it's primary `group` does **not** own the file
- salt is `root` and the `permissive_pki_access` config option is true and the `root` user is a member of the file owning group
- salt is **not** `root` and `others` do **not** have write access and `permissive_pki_access` is true and the owning group has write access
- salt is **not** `root` and `others` do **not** have write access and the owning `group` does **not** have write access

Which seems pretty random to me, especially since basically anything is allowed as long as salt runs as `root` and owns the file, even if anybody can write to the file.

### New Behavior
The function checks exactly that:
- `Others` must **not** have write access to the file.
- if the `group` has write access to the file: `permissive_pki_access` must be true
    - if salt is `root`: the `root` user must also be part of that `group`

### Tests written?
Yes

See #44175 for previous discussions on merging this into 2016.11.

### CODE REVIEW

File: conf/master [hunk 1]
```diff
@@ -337,7 +337,8 @@
 
 # If the autosign_file is specified, incoming keys specified in the
 # autosign_file will be automatically accepted. This is insecure.  Regular
-# expressions as well as globing lines are supported.
+# expressions as well as globing lines are supported. The file must be readonly
+# except for the owner. User permissive_pki_access to allow the group write access.
```
Comment (rallytime) [inline conf/master]: `User` should be `Use`.

File: conf/suse/master [hunk 2]
```diff
@@ -311,7 +311,8 @@ syndic_user: salt
 
 # If the autosign_file is specified, incoming keys specified in the
 # autosign_file will be automatically accepted. This is insecure.  Regular
-# expressions as well as globing lines are supported.
+# expressions as well as globing lines are supported. The file must be readonly
+# except for the owner. User permissive_pki_access to allow the group write access.
```
Comment (rallytime) [inline conf/suse/master]: `User` should be `Use`.

File: doc/ref/configuration/master.rst [hunk 3]
```diff
@@ -1275,6 +1275,9 @@ Default: ``not defined``
 If the ``autosign_file`` is specified incoming keys specified in the autosign_file
 will be automatically accepted. Matches will be searched for first by string
 comparison, then by globbing, then by full-string regex matching.
+The file must be readonly except for it's owner.
+If :conf_master:`permissive_pki_access` is ``True`` the owning group can also
+have write access, but if salt is running as ``root`` it must be a member of that group.
```
Comment (rallytime) [inline doc/ref/configuration/master.rst]: `salt` should be `Salt`.

**Review by rallytime — CHANGES_REQUESTED** [review]
@benediktwerner I had a couple of minor grammar fixes. Otherwise this looks good to me.

Thank you for adding these tests!

**Comment by benediktwerner** [comment]
Fixed.

**Review by rallytime — APPROVED** [review]

**Comment by cachedout** [comment]
@benediktwerner Can you please resolve the merge conflict here? Thanks.

**Comment by benediktwerner** [comment]
Done.

**Review by thatch45 — APPROVED** [review]

**Pull request merged**